### PR TITLE
fix: nft shape memory issue

### PR DESCRIPTION
**Pull request by BrianAmadori**
## What does this PR change?

- New criteria for limiting preview images for NFTs added: 500kb limit
- This limit is not applicable for gif NFTs.
- Added tests to ensure the criteria keeps working.
- Refactored part of the image loading code to keep it simple and cheaper to maintain. Now `GifPlayer` lives inside the `NFTAsset_Gif` instance, `IPromiseLike` abstraction was removed, factory was removed, many methods were renamed because their names didn't match their behaviours, promises lifecycle now are handled inside the `NFTAssetLoadHelper` class, etc.
- Fixed some weird behaviour where `FinishLoading` method was being called twice.

### CODE REVIEW

**Review by sandrade-dcl — APPROVED** [review]
The code looks good! I also did a general test around the world and I didn't find problems. Anyway, do you know any specific scene where we can go to test it?

**Review by pbosio — CHANGES_REQUESTED** [review]
HQ NFTs are broken in this branch, it fail trying to fetch the higher resolution images

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTShapeHQImageHandler.cs [hunk 1]
```diff
@@ -98,20 +100,21 @@ private void OnPlayerPositionChanged(Vector3 current, Vector3 prev)
...
             if (!isPlayerNear)
             {
                 RestorePreviewTextureIfInHQ();
             }
 
-            if (config.nftConfig.verbose)
+            if (VERBOSE)
             {
-                Debug.Log($"Player position relative to {config.nftInfo.name} is near? {isPlayerNear}");
+                //Debug.Log($"Player position relative to {hqImageConfig.name} is near? {isPlayerNear}");
```
Comment (AjimenezDCL) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTShapeHQImageHandler.cs:117]: Shall we remove this?

**Review by AjimenezDCL — APPROVED** [review]
Left a small comment, the NFT system is quite hard to follow, it took me a while to review this PR. In any case I didnt see issues, good job!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTAsset/NFTAssetLoadHelper.cs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+            yield return GetHeaders(url, headers, result => responseHeaders = result, (x) => headerRequestError = x);
+
+            if (!string.IsNullOrEmpty(headerRequestError))
+            {
+                OnFail?.Invoke(new Exception($"Error fetching headers! ({headerRequestError})"));
+                yield break;
+            }
+
+            Debug.Log("Loading preview... " + url);
```
Comment (pbosio) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTAsset/NFTAssetLoadHelper.cs:46]:
```suggestion
```

File: unity-renderer/Assets/Scripts/MainScripts/DCL/ServiceProviders/OpenSea/Requests/RequestBase.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@ public abstract class RequestBase<T>
         public event Action<RequestBase<T>> OnSuccess;
         public event Action<RequestBase<T>> OnFail;
 
-        public bool resolved { private set; get; }
+        public bool resolved { private set; get; }
```
Comment (pbosio) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/ServiceProviders/OpenSea/Requests/RequestBase.cs:10]:
```suggestion
        public bool resolved { private set; get; }
```

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTAsset/NFTAssetLoadHelper.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            string contentType = responseHeaders[CONTENT_TYPE];
+            long contentLength = long.Parse(responseHeaders[CONTENT_LENGTH]);
```
Comment (pbosio) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/LoadableShapes/NFTShape/NFTAsset/NFTAssetLoadHelper.cs:48]: could we use `TryParse` here? an issue on the server side might result on the response header not containing `Content-Length` it's not very common but it might be better to prevent exception here

**Review by pbosio — APPROVED** [review]
